File: sdk/javascript/src/nem/SharedKey.js
/* eslint-disable no-unused-vars */
import { KeyPair } from './KeyPair.js';
/* eslint-enable no-unused-vars */
import {
	/* eslint-disable no-unused-vars */
	PublicKey,
	/* eslint-enable no-unused-vars */
	SharedKey256
} from '../CryptoTypes.js';
import { deriveSharedKeyFactory, deriveSharedSecretFactory } from '../SharedKey.js';
import { keccak_256, keccak_512 } from '@noble/hashes/sha3';

const crypto_hash = (out, m, n) => {
	const hashBuilder = keccak_512.create();
	hashBuilder.update(m.subarray(0, n));
	const hash = hashBuilder.digest();

	for (let i = 0; i < out.length; ++i)
		out[i] = hash[i];

	return 0;
};
const deriveSharedSecretImpl = deriveSharedSecretFactory(crypto_hash);
const deriveSharedKeyImpl = deriveSharedKeyFactory('nem-nis1', crypto_hash);

/**
 * Derives shared key from key pair and other party's public key.
 * @param {KeyPair} keyPair Key pair.
 * @param {PublicKey} otherPublicKey Other party's public key.
 * @returns {SharedKey256} Shared encryption key.
 */
const deriveSharedKey = (keyPair, otherPublicKey) => {
	const reversedPrivateKeyBytes = new Uint8Array([...keyPair.privateKey.bytes]);
	reversedPrivateKeyBytes.reverse();

	return deriveSharedKeyImpl(reversedPrivateKeyBytes, otherPublicKey);
};

/**
 * Derives shared key from key pair, other party's public key and salt
 * @deprecated This is _old_ method of deriving shared key and should not be used in new applications.
 * @param {KeyPair} keyPair Key pair.
 * @param {PublicKey} otherPublicKey Other party's public key.
 * @param {Uint8Array} salt Random salt. Should be unique per every use.
 * @returns {SharedKey256} Shared encryption key.
 */
const deriveSharedKeyDeprecated = (keyPair, otherPublicKey, salt) => {
	if (SharedKey256.SIZE !== salt.length)
		throw Error('invalid salt');

	const reversedPrivateKeyBytes = new Uint8Array([...keyPair.privateKey.bytes]);
	reversedPrivateKeyBytes.reverse();

	const sharedSecret = deriveSharedSecretImpl(reversedPrivateKeyBytes, otherPublicKey);
	const sharedKeyBytes = new Uint8Array(SharedKey256.SIZE);
	for (let i = 0; SharedKey256.SIZE > i; ++i)
		sharedKeyBytes[i] = sharedSecret[i] ^ salt[i];

	return new SharedKey256(keccak_256(sharedKeyBytes));
};

export {
	deriveSharedKey,
	deriveSharedKeyDeprecated
};

File: sdk/cpp/symbolchain/nem/SharedKey.py
import sha3

from ..CryptoTypes import SharedKey256
from ..external import ed25519
from ..SharedKey import SharedKey as BasicSharedKey


class SharedKey(BasicSharedKey):
	@staticmethod
	def derive_shared_key(key_pair, other_public_key):
		"""Derives shared encryption key from private key and public key."""

		# note: key_pair.private_key is "unreversed", so need to reverse here again
		return BasicSharedKey._derive_shared_key(other_public_key.bytes, key_pair.private_key.bytes[::-1], b'nem-nis1', sha3.keccak_512)

	@staticmethod
	def derive_shared_key_deprecated(key_pair, other_public_key, salt):
		"""Derives shared encryption key from private key, public key and salt.

		note: this method uses _old_ method of deriving shared key and has been deprecated.
		"""

		# note: key_pair.private_key is "unreversed", so need to reverse here again
		shared_secret = ed25519.derive_shared_secret_unsafe(other_public_key.bytes, key_pair.private_key.bytes[::-1], sha3.keccak_512)
		key = bytes([shared_secret[i] ^ salt[i] for i in range(32)])
		return SharedKey256(sha3.keccak_256(key).digest())
